src/palace/manager/service/redis/redis.py

```diff
@@ -0,0 +1,154 @@
+    """
+    A mixin that with functions to check that the keys used in a redis command have the expected
+    prefix. This is useful for ensuring that keys are namespaced correctly in a multi-tenant
+    environment.
+
+    We use this mixin in our Redis and Pipeline classes.
+
+    Some inspiration for this was taken from Kombu's Redis class. See:
+        https://github.com/celery/kombu/pull/1349
+    """
+
+    class RedisCommandArgsBase(ABC):
+        def __init__(self, name: str):
+            self.name = name
+
+        @abstractmethod
+        def key_args(self, args: list[Any]) -> Sequence[str]:
+            """
+            Takes a list of arguments and returns a sequence of keys that should be checked for the
+            correct prefix.
+            """
+            ...
+
+    class RedisCommandArgs(RedisCommandArgsBase):
+        def __init__(self, name: str, *, args_start: int = 0, args_end: int | None = 1):
+            super().__init__(name)
+            self.args_start = args_start
+            self.args_end = args_end
+
+        def key_args(self, args: list[Any]) -> Sequence[str]:
+            return [str(arg) for arg in args[self.args_start : self.args_end]]
+
+    class RedisVariableCommandArgs(RedisCommandArgsBase):
+        def __init__(self, name: str, *, key_index: int = 0):
+            super().__init__(name)
+            self.key_index = key_index
+
+        def key_args(self, args: list[Any]) -> Sequence[str]:
+            keys = int(args[self.key_index])
+            args_start = self.key_index + 1
+            return [str(arg) for arg in args[args_start : args_start + keys]]
+
+    class RedisCommandNoArgs(RedisCommandArgsBase):
+        def key_args(self, args: list[Any]) -> Sequence[str]:
+            return []
+
+    _PREFIXED_COMMANDS = {
+        cmd.name: cmd
+        for cmd in [
+            RedisCommandNoArgs("SCRIPT LOAD"),
+            RedisCommandArgs("KEYS"),
+            RedisCommandArgs("GET"),
+            RedisCommandArgs("GETRANGE"),
+            RedisCommandArgs("SET"),
+            RedisCommandArgs("TTL"),
+            RedisCommandArgs("DEL", args_end=None),
+            RedisCommandArgs("MGET", args_end=None),
+            RedisCommandArgs("EXPIRETIME"),
+            RedisVariableCommandArgs("EVALSHA", key_index=1),
+        ]
+    }
+
+    def _check_prefix(self, *args: Any) -> None:
+        arg_list = list(args)
+        command = arg_list.pop(0)
+        cmd_args = self._PREFIXED_COMMANDS.get(command)
+        if cmd_args is not None:
+            for key in cmd_args.key_args(arg_list):
+                if not key.startswith(self._prefix):
+                    raise RedisValueError(
+                        f"Key {key} does not start with prefix {self._prefix}. Command {command} args: {arg_list}"
+                    )
+        else:
+            raise RedisValueError(
+                f"Command {command} is not checked for prefix. Args: {arg_list}"
+            )
```

Comment:
Minor: Swapping the `if` and `else` blocks would let us guard and simplify here:
```python
  if cmd_args is None:
      raise RedisValueError(
          f"Command {command} is not checked for prefix. Args: {arg_list}"
      )
  for key in cmd_args.key_args(arg_list):
      ...
```